### PR TITLE
C# language reference for records

### LINKED ISSUE #20585: Language Reference: Record types
There are a few tasks here:

- [x] Update the "class & structs" article to include a short introductory section on `record` types.
- [x] Add `with` to keywords
- [x] Add an article that describes `with`-ers.
- [x] Add a new language reference article for `records`. This should include information on the compiler generated members, and how to override them. In addition, it should explain inheritance rules for records, primary constructors, with-ers, and deconstruction methods.
- [x] Update the "Programming Guide->Types->Use and define types" article to mention records.
- [x] Update the "Programming Guide -> Classes and structs" with an article on records. The new article should explain records and when to use them over class types.

### PR DESCRIPTION
**Pull request by tdykstra**
Fixes #20585

### CODE REVIEW

File: docs/csharp/language-reference/builtin-types/record.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+### `PrintMembers` formatting in derived records
+
+the synthesized `PrintMembers` method of a derived record type calls the base implementation. The result is that all public properties and fields of both derived and base types are included in the `ToString` output, as shown in the following example:
+
+:::code language="csharp" source="snippets/shared/RecordType.cs" id="ToStringInheritance":::
+
+You can provide your own implementation of the `PrintMembers` method. If you do that, use the following signature:
+
+* For a `sealed` record: `private bool PrintMembers(StringBuilder builder)`;
```
Comment (BillWagner) [inline docs/csharp/language-reference/builtin-types/record.md:171]: This needs to be clarified a bit:

```suggestion
* For a `sealed` record that derives from `object`: `private bool PrintMembers(StringBuilder builder)`;
```

IIRC, a sealed record that derives from another record would have the signature `protected sealed override bool PrintMembers(StringBuilder builder);`

File: docs/csharp/language-reference/builtin-types/record.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+* For a record that isn't `sealed` and derives from object (doesn't declare a base record): `protected virtual bool PrintMembers(StringBuilder builder);`
+* For a record that isn't `sealed` and derives from another record: `protected override bool PrintMembers(StringBuilder builder);`
+
+Here is an example of code that replaces the synthesized `PrintMembers` methods, one for a record type that derives from object, and one for a record type that derives from another record:
+
+:::code language="csharp" source="snippets/shared/RecordType.cs" id="PrintMembersImplementation":::
+
+### Deconstructor behavior in derived records
+
+The `Deconstructor` method of a derived record returns the values of all positional properties of the compile-time type. If the variable type is a base record, only the base record properties are deconstructed unless the object is cast to the derived type. The following example demonstrates calling a deconstructor on a derived record.
```
Comment (BillWagner) [inline docs/csharp/language-reference/builtin-types/record.md:181]: nit on the name:

```suggestion
The `Deconstruct` method of a derived record returns the values of all positional properties of the compile-time type. If the variable type is a base record, only the base record properties are deconstructed unless the object is cast to the derived type. The following example demonstrates calling a deconstructor on a derived record.
```

**Review by BillWagner — COMMENTED** [review]
This looks really good @tdykstra 

I had a couple nits and one or two other suggestions.

Once we can see the preview site, I'll give it another look.

File: docs/csharp/language-reference/builtin-types/record.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+Beginning with C# 9, you use the `record` keyword to define a [reference type](reference-types.md) that provides built-in functionality for encapsulating data. You can create record types with immutable properties by using positional parameters or standard property syntax:
+
+:::code language="csharp" source="snippets/shared/RecordType.cs" id="PositionalRecord":::
+:::code language="csharp" source="snippets/shared/RecordType.cs" id="ImmutableRecord":::
+
+You can also create record types with mutable properties and fields:
+
+:::code language="csharp" source="snippets/shared/RecordType.cs" id="MutableRecord":::
+
+While records can be mutable, they are primarily intended for supporting immutable data models. The record type offers the following features:
```
[the remark this answers is not in the document]
Comment (tdykstra) [reply docs/csharp/language-reference/builtin-types/record.md:22]: That statement was adapted from "While records can be mutable, they are primarily built for better supporting immutable data models." in [Mads' blog]((https://devblogs.microsoft.com/dotnet/c-9-0-on-the-record/#records). 😄

[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/language-reference/builtin-types/record.md:22]: Well then, let's just :shipit:

File: docs/csharp/language-reference/builtin-types/record.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+* [Concise syntax for creating a reference type with immutable properties](#positional-syntax-for-property-definition)
+* Built-in behavior useful for a data-centric reference type:
+  * [Value equality](#value-equality)
+  * [Concise syntax for nondestructive mutation](#nondestructive-mutation)
```
[the remark this answers is not in the document]
Comment (tdykstra) [reply docs/csharp/language-reference/builtin-types/record.md:29]: I found that it works with non-positional records.  For example, the following code works as expected:

```
            public record Point
            {
                public int X { get; init; }
                public int Y { get; init; }
                public int Zbase { get; init; }
            };
            public record NamedPoint : Point
            {
                public string Name { get; init; }
                public int Zderived { get; set; }
            };

            public static void Main()
            {
                Point p1 = new NamedPoint{ Name ="A", X = 1, Y = 2, Zbase = 3, Zderived = 4 };

                Point p2 = p1 with { X = 5, Y = 6, Zbase = 7 }; // Can't set Name or Zderived
                Console.WriteLine(p2 is NamedPoint);  // output: True
                Console.WriteLine(p2);
                // output: NamedPoint { X = 5, Y = 6, Zbase = 7, Name = A, Zderived = 4 }

                Point p3 = (NamedPoint)p1 with { Name = "B", X = 5, Y = 6, Zbase = 7, Zderived = 8 };
                Console.WriteLine(p3);
                // output: NamedPoint { X = 5, Y = 6, Zbase = 7, Name = B, Zderived = 8 }
            }
```

[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/language-reference/builtin-types/record.md:29]: That is cool. I was going from memory, and happy to be wrong.

File: docs/csharp/language-reference/builtin-types/record.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+To implement value equality, the compiler synthesizes the following methods:
+
+* An override of <xref:System.Object.Equals(System.Object)?displayProperty=nameWithType>.
+
+  This method is used as the basis for the <xref:System.Object.Equals(System.Object,System.Object)?displayProperty=nameWithType> static method when both parameters are non-null.
+
+* A virtual `Equals` method whose parameter is the record type.
+
+  The record type implements <xref:System.IEquatable%601>
```
Comment (tdykstra) [inline docs/csharp/language-reference/builtin-types/record.md:88]:
```suggestion
  The record type implements <xref:System.IEquatable%601> where `T` is the record type.
```

File: docs/csharp/language-reference/builtin-types/record.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+The features unique to record types are implemented by compiler-synthesized methods, and none of these methods compromises immutability by modifying object state.
+
+## Value equality
+
+Value equality means that two variables of a record type are equal if the types match and all property and field values match. For other reference types, equality means identity. That is, two variables of a reference type are equal if they refer to the same object. The following example illustrates value equality of record types:
+
+:::code language="csharp" source="snippets/shared/RecordType.cs" id="Equality":::
+
+To implement value equality, the compiler synthesizes the following methods:
```
[the remark this answers is not in the document]
Comment (tdykstra) [reply docs/csharp/language-reference/builtin-types/record.md:80]: Do you mean `IEquatable<T>`?

File: docs/csharp/language-reference/builtin-types/record.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+:::code language="csharp" source="snippets/shared/RecordType.cs" id="PositionalRecord":::
+:::code language="csharp" source="snippets/shared/RecordType.cs" id="ImmutableRecord":::
+
+You can also create record types with mutable properties and fields:
+
+:::code language="csharp" source="snippets/shared/RecordType.cs" id="MutableRecord":::
+
+Positional records create immutable data models. You can add additional properties that are mutable to positional records. You can control the mutability of any property in non-positional record types.
+
+ The record type offers the following features:
```
Comment (tdykstra) [inline docs/csharp/language-reference/builtin-types/record.md:24]:
```suggestion
While records can be mutable, they are primarily intended for supporting immutable data models. The record type offers the following features:
```

File: docs/csharp/language-reference/builtin-types/record.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/language-reference/builtin-types/record.md:78]: yes.

File: docs/csharp/language-reference/builtin-types/record.md [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+the synthesized `PrintMembers` method of a derived record type calls the base implementation. The result is that all public properties and fields of both derived and base types are included in the `ToString` output, as shown in the following example:
+
+:::code language="csharp" source="snippets/shared/RecordType.cs" id="ToStringInheritance":::
+
+You can provide your own implementation of the `PrintMembers` method. If you do that, use the following signature:
+
+* For a `sealed` record that derives from `object` (doesn't declare a base record): `private bool PrintMembers(StringBuilder builder)`;
+* For a `sealed` record that derives from another record: `protected sealed override bool PrintMembers(StringBuilder builder)`;
+* For a record that isn't `sealed` and derives from object (doesn't declare a base record): `protected virtual bool PrintMembers(StringBuilder builder);`
```
Comment (tdykstra) [inline docs/csharp/language-reference/builtin-types/record.md:173]:
```suggestion
* For a record that isn't `sealed` and derives from object: `protected virtual bool PrintMembers(StringBuilder builder);`
```

File: docs/csharp/language-reference/builtin-types/record.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (tdykstra) [reply docs/csharp/language-reference/builtin-types/record.md:78]: I have that below as an explanatory comment to the `* A virtual Equals method whose parameter is the record type.` bullet point -- should it be a separate bullet point?

[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/language-reference/builtin-types/record.md:78]: good point. They are related. Maybe combine them into one bullet point? I think the interface should be mentioned, which introduces the second `Equals` method.

File: docs/csharp/language-reference/builtin-types/record.md [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+### `with` expressions in derived records
+
+Because the synthesized clone method uses a [covariant return type](~/_csharplang/proposals/csharp-9.0/covariant-returns.md), the result of a `with` expression has the same run-time type as the expression's operand. All properties of the run-time type get copied, but you can only set properties of the compile-time type, as the following example shows:
+
+:::code language="csharp" source="snippets/shared/RecordType.cs" id="WithExpressionInheritance":::
+
+### `PrintMembers` formatting in derived records
+
+the synthesized `PrintMembers` method of a derived record type calls the base implementation. The result is that all public properties and fields of both derived and base types are included in the `ToString` output, as shown in the following example:
```
Comment (pkulikov) [inline docs/csharp/language-reference/builtin-types/record.md:167]:
```suggestion
The synthesized `PrintMembers` method of a derived record type calls the base implementation. The result is that all public properties and fields of both derived and base types are included in the `ToString` output, as shown in the following example:
```

Comment (tdykstra) [reply docs/csharp/language-reference/builtin-types/record.md]: Thanks @pkulikov

**Comment by tdykstra** [comment]
Closing and reopening to trigger build

**Review by BillWagner — APPROVED** [review]
This is awesome stuff @tdykstra 

I really like the way it came together.

You can :shipit: when ready.

**Pull request merged**